Drop databases after unit tests if the database instance is not running in docker

zcmander opened the pull request:
## Hey, I just made a Pull Request!

This PR extends backend-test-utils to drop databases after tests has run if the database instance is not running in docker. This is to make sure that we don't inadvertently fill the database instance with bunch of test databases. It is assumed that if you give a connection string to external database then the database is more "permanent" that the one that has been launched in a container just for the tests.

Relates to the discussion here: #21619

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

benjdlambert commented:
Gonna hand this over to you @freben as I know we discussed this before, think you have more context on this :)

freben commented:
Note the build errors about using `const` and running prettier on changed files (you may want to configure your editor to auto-format on save).

freben commented on packages/backend-common/src/database/DatabaseManager.ts:
```diff
@@ -111,6 +111,30 @@ export class DatabaseManager {
+    const pluginConfig = new ConfigReader(
+      this.getConfigForPlugin(pluginId) as JsonObject,
+    );
+    const client = createDatabaseClient(pluginConfig, {}, deps);
```
Hm, the client that's constructed when _creating_ databases is unique, database specific, and has higher privileges; [see the code here](https://github.com/backstage/backstage/blob/bb6bce457ed069a38ef8d16ef38602972c7735c5/packages/backend-common/src/database/connectors/postgres.ts#L125) and note that that's in the postgres connector, not generic code. The one [in the mysql connector](https://github.com/backstage/backstage/blob/bb6bce457ed069a38ef8d16ef38602972c7735c5/packages/backend-common/src/database/connectors/mysql.ts#L150) is, again, different. You would need to add a dropDatabase and dropSchema into the connectors system for this (where only the pg connector actually implements the dropSchema).

Hm. Let me back up a bit.

All of this makes me ponder if it's unfortunate that the TestDatabases use this class at all. Deleting things just isn't part of what it was meant to do, and it kind of shows at this point. TestDatabases already knows that the BASE credentials it has been given are enough to do admin actions, because of how it's designed; it doesn't need to elevate itself. DatabaseManager which is in the production path does not have that luxury. Now the proposed code in here "leaks" that difference into the DatabaseManager class. The TestDatabases don't care at all about plugin division modes either - that's only relevant in production.

OK, in summary, I think this class should actually be left unchanged. Is there some way of ONLY putting the deletion code inside TestDatabases? Then we can as a separate task in the future see if we can just plain detach TestDatabases from DatabaseManager.

zcmander replied:
Yup, I totally agree with that. I had hard time understanding the responsibility of the database manager, so thanks for clarifying that. I'll make the proposed changes later.

zcmander replied:
Done.

freben commented on packages/backend-test-utils/src/database/types.ts:
```diff
@@ -45,6 +45,8 @@ export type Instance = {
   stopContainer?: () => Promise<void>;
   databaseManager: DatabaseManager;
   connections: Array<Knex>;
+  dropDatabasesInShutdown: boolean;
```
```suggestion
  dropDatabase?: () => Promise<void>;
```

seems like the better contract, then each database driver can both choose whether to return it, and how it's implemented.

Then the code above would be exactly like the stopContainer code

```ts
      try {
        await dropDatabase?.();
      } catch (error) {
        console.warn(`TestDatabases: Failed to drop database`, {
          databaseManager,
          error,
        });
      }
```

zcmander replied:
Modified the code to use callback-style contract.

zcmander commented on packages/backend-common/src/database/DatabaseManager.ts:
```diff
@@ -111,6 +111,30 @@ export class DatabaseManager {
+    const client = createDatabaseClient(pluginConfig, {}, deps);
+
+    if (this.getPluginDivisionModeConfig() === 'schema') {
+      await client.raw(`DROP SCHEMA ${pluginId}`);
```
Use parameter binding: https://knexjs.org/guide/raw.html

zcmander commented:
Requesting re-review @freben . I could not put *ALL* code in the `TestDatabase` class but now `DatabaseManager` remains unchanged.

I made the suggested improvements:
 * moved drop database functionality to mysql and postgresql drivers
 * removed all modifications from `DatabaseManager`
 * replaced boolean-flag with callback-style hook to easily separate container-based and envvar-based databases from each other.

zcmander commented:
not stale, waiting for review :/

zcmander commented:
not stale, just waiting for review :/

freben reviewed (approved):
I think this is ready to go!
